### DISCUSSION TITLE
Seemingly valid openapi schema  that Swagger can't parse (Body  annotated dict  with datetime key)

### DISCUSSION THREAD

**Discussion by chaen** [fastapi/fastapi#12984] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Body
from typing import Annotated
from datetime import datetime
app = FastAPI()
app.openapi_version  = "3.0.2"




@app.patch("/working")
async def working(body:Annotated[dict[str,str], Body()]):
    return {"message": f"Hello World: {body}"}


@app.patch("/failing")
async def failing(body:Annotated[dict[datetime,str], Body()]):
    return {"message": f"Hello World: {body}"}
```

### Description

* Download the `openapi.json` file (`curl -O http://localhost:8000/openapi.json`)
* Paste it into the swagger-ui editor (https://editor.swagger.io/)
* Observe the error

```
Errors
 
Structural error at paths./failing.patch.requestBody.content.application/json.schema
should NOT have additional properties
additionalProperty: propertyNames
Jump to line 38
 ```

<details>
<summary>For completeness, here is the json generated</summary>

```json
{
    "openapi": "3.0.2",
    "info": {
        "title": "FastAPI",
        "version": "0.1.0"
    },
    "paths": {
        "/working": {
            "patch": {
                "summary": "Working",
                "operationId": "working_working_patch",
                "requestBody": {
                    "content": {
                        "application/json": {
                            "schema": {
                                "additionalProperties": {
                                    "type": "string"
                                },
                                "type": "object",
                                "title": "Body"
                            }
                        }
                    },
                    "required": true
                },
                "responses": {
                    "200": {
                        "description": "Successful Response",
                        "content": {
                            "application/json": {
                                "schema": {}
                            }
                        }
                    },
                    "422": {
                        "description": "Validation Error",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/HTTPValidationError"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/failing": {
            "patch": {
                "summary": "Failing",
                "operationId": "failing_failing_patch",
                "requestBody": {
                    "content": {
                        "application/json": {
                            "schema": {
                                "additionalProperties": {
                                    "type": "string"
                                },
                                "propertyNames": {
                                    "format": "date-time"
                                },
                                "type": "object",
                                "title": "Body"
                            }
                        }
                    },
                    "required": true
                },
                "responses": {
                    "200": {
                        "description": "Successful Response",
                        "content": {
                            "application/json": {
                                "schema": {}
                            }
                        }
                    },
                    "422": {
                        "description": "Validation Error",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/HTTPValidationError"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "HTTPValidationError": {
                "properties": {
                    "detail": {
                        "items": {
                            "$ref": "#/components/schemas/ValidationError"
                        },
                        "type": "array",
                        "title": "Detail"
                    }
                },
                "type": "object",
                "title": "HTTPValidationError"
            },
            "ValidationError": {
                "properties": {
                    "loc": {
                        "items": {
                            "anyOf": [
                                {
                                    "type": "string"
                                },
                                {
                                    "type": "integer"
                                }
                            ]
                        },
                        "type": "array",
                        "title": "Location"
                    },
                    "msg": {
                        "type": "string",
                        "title": "Message"
                    },
                    "type": {
                        "type": "string",
                        "title": "Error Type"
                    }
                },
                "type": "object",
                "required": [
                    "loc",
                    "msg",
                    "type"
                ],
                "title": "ValidationError"
            }
        }
    }
}
```
</details>

Notice that it works fine for `dict[str,str]`

From the little understanding I have, the schema is actually a valid JSON 2020-12/OpenAPI 3.1 (as implemented in FastAPI https://github.com/fastapi/fastapi/pull/9782/commits/7dad5a820bfe99e49a6cfaefde537e09644c2c2b)  but multiple tools fail at parsing it properly, for example:
* [swagger](https://github.com/swagger-api/swagger-ui/issues/8913) 
* [autorest](https://github.com/Azure/autorest/issues/3727)

Can anybody help me confirm that the problem is indeed what I have understood (i.e. generation is correct, but other tools fail to read it properly) or is there a bug in how it is generated (as it works for `str`, which is a basic type, so maybe different). 
If so many tools fail to ingest the json file produced by `FastAPI`, would it be worth considering an option to generate older schema ? It seems to me that changing the `openapi_version` as I did in my test script only changes the version number, not the actual schema. 

Thanks  a lot ! 
Chris

### Operating System

Linux

### FastAPI Version

0.115.5

### Pydantic Version

2.10.1

### Python Version

3.11.10

**Comment by agammann** [comment]
`app.openapi_version = "3.0.2"` changes the declared version; it does not convert the generated schema to OpenAPI 3.0. [FastAPI 0.115.5 documents that limitation in the attribute itself](https://github.com/fastapi/fastapi/blob/0.115.5/fastapi/applications.py#L842-L859).

I checked your two routes locally with FastAPI 0.115.5 and Pydantic 2.10.1 on Python 3.12.14 / Windows. The datetime-key schema contains `"propertyNames": {"format": "date-time"}`. Using `openapi-spec-validator` 0.9.0:

- The default OpenAPI 3.1.0 document passes validation.
- Setting the version to 3.0.2, clearing `app.openapi_schema`, and regenerating leaves the route schemas unchanged; the 3.0 validator rejects the document.

For this example, remove the version override and use OpenAPI 3.1-compatible tooling. Relabelling the document does not provide 3.0 compatibility. This checks schema generation and validation; it does not establish compatibility with a particular client generator.

Edited to replace the untested workaround suggestions with these measured results.
